src/main/java/io/cdap/plugin/google/sheets/sink/GoogleSheetsRecordWriter.java

```diff
@@ -0,0 +1,373 @@
+/*
+ * Copyright © 2019 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.plugin.google.sheets.sink;
+
+import com.github.rholder.retry.RetryException;
+import com.google.api.services.sheets.v4.model.Request;
+import com.google.api.services.sheets.v4.model.SheetProperties;
+import com.google.api.services.sheets.v4.model.Spreadsheet;
+import io.cdap.plugin.google.drive.sink.GoogleDriveOutputFormatProvider;
+import io.cdap.plugin.google.drive.sink.GoogleDriveSinkClient;
+import io.cdap.plugin.google.sheets.sink.utils.DimensionType;
+import io.cdap.plugin.google.sheets.sink.utils.FlatternedRowsRecord;
+import io.cdap.plugin.google.sheets.sink.utils.FlatternedRowsRequest;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.io.NullWritable;
+import org.apache.hadoop.mapreduce.RecordWriter;
+import org.apache.hadoop.mapreduce.TaskAttemptContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Writes {@link FlatternedRowsRecord} records to Google drive via {@link GoogleDriveSinkClient}.
```

Comment:
Good big javadoc. How everything is merged, how queue works, why we need it etc.

Reply:
commented